### PR TITLE
DevWs for YookSmart containing containing ZigBee Window Shade Battery

### PR DESCRIPTION
**Pull request by JeremeLau**

### CODE REVIEW

File: devicetypes/smartthings/zigbee-window-shade-battery.src/zigbee-window-shade-battery.groovy [hunk 1]
```diff
@@ -22,7 +21,6 @@ metadata {
 		capability "Configuration"
 		capability "Refresh"
 		capability "Window Shade"
-		capability "Window Shade Level"
```
Comment (KKlimczukS) [inline devicetypes/smartthings/zigbee-window-shade-battery.src/zigbee-window-shade-battery.groovy]: Why did You remove this capability ?

File: devicetypes/smartthings/zigbee-window-shade-battery.src/zigbee-window-shade-battery.groovy [hunk 2]
```diff
@@ -34,28 +32,22 @@ metadata {
 		fingerprint manufacturer: "IKEA of Sweden", model: "FYRTUR block-out roller blind", deviceJoinName: "IKEA Window Treatment" // raw description 01 0104 0202 01 09 0000 0001 0003 0004 0005 0020 0102 1000 FC7C 02 0019 1000 //IKEA FYRTUR Blinds
 
 		// Yookee yooksmart
-		fingerprint inClusters: "0000,0001,0003,0004,0005,0102", outClusters: "0019", manufacturer: "Yookee", model: "D10110", deviceJoinName: "Yookee Window Treatment"
-		fingerprint inClusters: "0000,0001,0003,0004,0005,0102", outClusters: "0019", manufacturer: "yooksmart", model: "D10110", deviceJoinName: "yooksmart Window Treatment"
-        
-		// SMARTWINGS
-		fingerprint inClusters: "0000,0001,0003,0004,0005,0102", outClusters: "0019", manufacturer: "Smartwings", model: "WM25/L-Z", deviceJoinName: "Smartwings Window Treatment"
```
Comment (KKlimczukS) [inline devicetypes/smartthings/zigbee-window-shade-battery.src/zigbee-window-shade-battery.groovy:41]: Why did You remove this fingerprint ?

File: devicetypes/smartthings/zigbee-window-shade-battery.src/zigbee-window-shade-battery.groovy [hunk 3]
```diff
@@ -34,28 +32,22 @@ metadata {
...
+		multiAttributeTile(name:"windowShade", type: "generic", width: 6, height: 4) {
 			tileAttribute("device.windowShade", key: "PRIMARY_CONTROL") {
-				attributeState "open", label: 'Open', action: "close", icon: "st.shades.shade-open", backgroundColor: "#00A0DC", nextState: "closing"
-				attributeState "closed", label: 'Closed', action: "open", icon: "st.shades.shade-closed", backgroundColor: "#ffffff", nextState: "opening"
-				attributeState "partially open", label: 'Partially open', action: "close", icon: "st.shades.shade-open", backgroundColor: "#00A0DC", nextState: "closing"
-				attributeState "opening", label: 'Opening', action: "pause", icon: "st.shades.shade-opening", backgroundColor: "#00A0DC", nextState: "partially open"
-				attributeState "closing", label: 'Closing', action: "pause", icon: "st.shades.shade-closing", backgroundColor: "#ffffff", nextState: "partially open"
-			}
-			tileAttribute ("device.windowShadeLevel", key: "SLIDER_CONTROL") {
-				attributeState "shadeLevel", action:"setShadeLevel"
```
Comment (KKlimczukS) [inline devicetypes/smartthings/zigbee-window-shade-battery.src/zigbee-window-shade-battery.groovy]: Why did You change this tileAttribute ?

File: devicetypes/smartthings/zigbee-window-shade-battery.src/zigbee-window-shade-battery.groovy [hunk 4]
```diff
@@ -34,28 +32,22 @@ metadata {
...
-		fingerprint inClusters: "0000,0001,0003,0004,0005,0102", outClusters: "0019", manufacturer: "Smartwings", model: "WM25/L-Z", deviceJoinName: "Smartwings Window Treatment"
+		fingerprint manufacturer: "Yookee", model: "D10110", deviceJoinName: "Yookee Window Treatment"
+		fingerprint manufacturer: "yooksmart", model: "D10110", deviceJoinName: "yooksmart Window Treatment"
 	}
 
 	preferences {
 		input "preset", "number", title: "Preset position", description: "Set the window shade preset position", defaultValue: 50, range: "1..100", required: false, displayDuringSetup: false
 	}
 
 	tiles(scale: 2) {
```
Comment (KKlimczukS) [inline devicetypes/smartthings/zigbee-window-shade-battery.src/zigbee-window-shade-battery.groovy]: tiles are not supported in the new ST app, so there is no need to change anything here.

File: devicetypes/smartthings/zigbee-window-shade-battery.src/zigbee-window-shade-battery.groovy [hunk 5]
```diff
@@ -100,31 +98,22 @@ private List<Map> collectAttributes(Map descMap) {
...
 }
 
 // Parse incoming device messages to generate events
 def parse(String description) {
 	log.debug "description:- ${description}"
-
-	if (device.currentValue("shadeLevel") == null && device.currentValue("level") != null) {
-		sendEvent(name: "shadeLevel", value: device.currentValue("level"), unit: "%")
-	}
-
```
Comment (KKlimczukS) [inline devicetypes/smartthings/zigbee-window-shade-battery.src/zigbee-window-shade-battery.groovy]: Why did You remove this ?

File: devicetypes/smartthings/zigbee-window-shade-battery.src/zigbee-window-shade-battery.groovy [hunk 6]
```diff
@@ -138,23 +127,18 @@ def parse(String description) {
...
 			def batteryLevel = zigbee.convertHexToInt(descMap.value)
-
 			batteryPercentageEventHandler(batteryLevel)
 		}
 	}
 }
 
 def levelEventHandler(currentLevel) {
-	def lastLevel = device.currentState("shadeLevel") ? device.currentValue("shadeLevel") : device.currentValue("level") // Try shadeLevel, if not use level and pass to logic below
-
```
Comment (KKlimczukS) [inline devicetypes/smartthings/zigbee-window-shade-battery.src/zigbee-window-shade-battery.groovy]: Why did You change that ?

**Review by KKlimczukS — CHANGES_REQUESTED** [review]
There are too many changes because of removing window "Shade Level capability".

File: devicetypes/smartthings/zigbee-window-shade-battery.src/zigbee-window-shade-battery.groovy [hunk 7]
```diff
@@ -169,56 +153,47 @@ def levelEventHandler(currentLevel) {
 }
 
 def updateFinalState() {
-	def level = device.currentValue("shadeLevel")
+	def level = device.currentValue("level")
```
Comment (dkirker) [inline devicetypes/smartthings/zigbee-window-shade-battery.src/zigbee-window-shade-battery.groovy:156]: This should be using `shadeLevel`.

File: devicetypes/smartthings/zigbee-window-shade-battery.src/zigbee-window-shade-battery.groovy [hunk 8]
```diff
@@ -169,56 +153,47 @@ def levelEventHandler(currentLevel) {
...
-
 	if (level > 0 && level < 100) {
 		sendEvent(name: "windowShade", value: "partially open")
 	}
 }
 
 def batteryPercentageEventHandler(batteryLevel) {
+	log.debug "batteryLevel: ${batteryLevel}"
 	if (batteryLevel != null) {
+    	if (isYooksmartOrYookee()) {
```
Comment (dkirker) [inline devicetypes/smartthings/zigbee-window-shade-battery.src/zigbee-window-shade-battery.groovy]: Please indent this if statement by two tabs (use the tab character, not spaces).

**Review by dkirker — CHANGES_REQUESTED** [review]
Please only add the changes to the fingerprint and this code to the battery handler with proper indenting:
```
    	if (isYooksmartOrYookee()) {
        	batteryLevel = batteryLevel >> 1
        }
```
Please do not change any of the other code unless it is necessary.

**Comment by dkirker** [comment]
I deleted some inline code comments as most of the inline code should NOT be changed. The only thing that should be changed is the addition of the if statement to the battery handler and possibly the modification of ONLY the Yooksmart and Yookee fingerprints.

File: devicetypes/smartthings/zigbee-window-shade-battery.src/zigbee-window-shade-battery.groovy [hunk 9]
```diff
@@ -34,9 +34,9 @@ metadata {
 		fingerprint manufacturer: "IKEA of Sweden", model: "FYRTUR block-out roller blind", deviceJoinName: "IKEA Window Treatment" // raw description 01 0104 0202 01 09 0000 0001 0003 0004 0005 0020 0102 1000 FC7C 02 0019 1000 //IKEA FYRTUR Blinds
 
 		// Yookee yooksmart
-		fingerprint inClusters: "0000,0001,0003,0004,0005,0102", outClusters: "0019", manufacturer: "Yookee", model: "D10110", deviceJoinName: "Yookee Window Treatment"
-		fingerprint inClusters: "0000,0001,0003,0004,0005,0102", outClusters: "0019", manufacturer: "yooksmart", model: "D10110", deviceJoinName: "yooksmart Window Treatment"
-        
+		fingerprint manufacturer: "Yookee", model: "D10110", deviceJoinName: "Yookee Window Treatment"
+		fingerprint manufacturer: "yooksmart", model: "D10110", deviceJoinName: "yooksmart Window Treatment"
```
Comment (KKlimczukS) [inline devicetypes/smartthings/zigbee-window-shade-battery.src/zigbee-window-shade-battery.groovy]: Please, could You add the raw description in the comment after each fingerprint ?

**Review by KKlimczukS — CHANGES_REQUESTED** [review]

**Review by KKlimczukS — DISMISSED** [review]

**Comment by KKlimczukS** [comment]
LGTM
@dkirker

File: devicetypes/smartthings/zigbee-window-shade-battery.src/zigbee-window-shade-battery.groovy [hunk 10]
```diff
@@ -149,7 +155,7 @@ def levelEventHandler(currentLevel) {
 
 	log.debug "levelEventHandle - currentLevel: ${currentLevel} lastLevel: ${lastLevel}"
 
-	if (lastLevel == "undefined" || currentLevel == lastLevel) { //Ignore invalid reports
```
Comment (dkirker) [inline devicetypes/smartthings/zigbee-window-shade-battery.src/zigbee-window-shade-battery.groovy:152]: @KKlimczukS @JeremeLau I don't think that we want to remove this.

File: devicetypes/smartthings/zigbee-window-shade-battery.src/zigbee-window-shade-battery.groovy [hunk 11]
```diff
@@ -67,9 +67,15 @@ metadata {
 		standardTile("refresh", "device.refresh", inactiveLabel: false, decoration: "flat", width: 2, height: 1) {
 			state "default", label:"", action:"refresh.refresh", icon:"st.secondary.refresh"
 		}
+		valueTile("shadeLevel", "device.level", width: 4, height: 1) {
```
Comment (dkirker) [inline devicetypes/smartthings/zigbee-window-shade-battery.src/zigbee-window-shade-battery.groovy]: While the tiles are not used any more, for the sake of cleanliness, let's please not add this and below to the PR.

File: devicetypes/smartthings/zigbee-window-shade-battery.src/zigbee-window-shade-battery.groovy [hunk 12]
```diff
@@ -67,9 +67,15 @@ metadata {
...
+			state "level", label: 'Shade is ${currentValue}% up', defaultState: true
+		}
 		valueTile("batteryLevel", "device.battery", width: 2, height: 2) {
 			state "battery", label:'${currentValue}% battery', unit:""
 		}
+		controlTile("levelSliderControl", "device.level", "slider", width:2, height: 1, inactiveLabel: false) {
```
Comment (dkirker) [inline devicetypes/smartthings/zigbee-window-shade-battery.src/zigbee-window-shade-battery.groovy]: While the tiles are not used any more, for the sake of cleanliness, let's please not add this and above to the PR.

File: devicetypes/smartthings/zigbee-window-shade-battery.src/zigbee-window-shade-battery.groovy (hunk 10 above)
Comment (dkirker) [inline devicetypes/smartthings/zigbee-window-shade-battery.src/zigbee-window-shade-battery.groovy]: @KKlimczukS  @JeremeLau I don't think that we want to remove this.

**Review by dkirker — CHANGES_REQUESTED** [review]
Still a couple more things to address.

**Review by KKlimczukS — DISMISSED** [review]

**Comment by KKlimczukS** [comment]
@dkirker

File: devicetypes/smartthings/zigbee-window-shade-battery.src/zigbee-window-shade-battery.groovy (hunk 10 above)
Comment (dkirker) [reply devicetypes/smartthings/zigbee-window-shade-battery.src/zigbee-window-shade-battery.groovy]: @JeremeLau Please add the `|| currentLevel == lastLevel` back in this line. If there is a reason that you are removing it, please explain here. Thanks

**Review by dkirker — CHANGES_REQUESTED** [review]
There is still line 152 that I don't think needs editing. If there is a reason why you want to remove it please explain why.

**Comment by JeremeLau** [comment]
> There is still line 152 that I don't think needs editing. If there is a reason why you want to remove it please explain why.

I have explained many times in other issues, cause it will block the UI rendering while someone try to re-open the blind when it is already in 100% with network error.

**Comment by JeremeLau** [comment]
> There is still line 152 that I don't think needs editing. If there is a reason why you want to remove it please explain why.

I have explained many times in other issues, cause it will block the UI rendering while someone try to re-open the blind when it is already in 100% with network error.

**Review by KKlimczukS — APPROVED** [review]
lgtm
@dkirker

**Review by dkirker — APPROVED** [review]

**Pull request merged**